### PR TITLE
Document Structure

### PR DESCRIPTION
**Pull request by graemenail**
### Description
This PR adds an outline of the code structure. 

List of changes:
- Adds ./doc/structure.md

Added dependencies: none

### How to test
Run `make html` after installing dependencies in `docs/`, or see documentation artifact generated by this PR for the output.

### Checklist

- [ ] I have tested the code manually
- [ ] I have run regression tests
- [ ] I have read and followed CONTRIBUTING.md
- [ ] I have updated CHANGELOG.md

### CODE REVIEW

File: doc/structure.md [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+```
+
+Marian implements a reverse-mode auto-differentiation computation graph. The relevant components reside in these subdirectories. The `graph` subdirectory concerns the structure of the graph, its nodes: operators, parameters and constants, as well as how to traverse it, both forwards and backwards. Moreover, it defines the APIs for operations that the graph is able to perform.
+
+The `tensors` and `functional` subdirectories contain the implementation of operations for the graph.
+
+One component of the `functional` subdirectory describes how functions operate on the underlying data types. This is a combination of standard operations on fundamental types, and SIMD intrinsics on extended types where available. The `functional` namespace also provides useful abstractions that enable generic formulas to be written. It defines variable-like objects `_1,_2`, such that `_1 * cos(_2)` represents the product of the argument at index 1 with the cosine of the argument at index 2.
+
+The `tensor` subdirectory contains the definition of a tensor object. In Marian, a tensor is a piece of memory which is ascribed a shape and type which is associated with a backend (the compute device).
```
Comment (snukky) [inline doc/structure.md:64]:
```suggestion
The `tensors` subdirectory contains the definition of a tensor object. In Marian, a tensor is a piece of memory which is ascribed a shape and type which is associated with a backend (the compute device).
```

File: doc/structure.md [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+
+Beyond these areas, this folder also contains metadata, such as the program version, list of contributors, and the build flags used to compile it.
+
+
+## External Libraries
+```
+marian/src
+└── 3rd_party
+  ```
+Many of the external libraries that marian depends on are contained in `3rd_party`.
```
Comment (snukky) [inline doc/structure.md:116]:
```suggestion
Many of the external libraries that Marian depends on are contained in `3rd_party`.
```

**Review by snukky — APPROVED** [review]
Thank you, @graemenail.

**Comment by graemenail** [comment]
I've addressed the comments in 65eccfc594ba971e7c75afb22038e9fa5b872e8e.

Thanks also for the presentation you pointed me to @snukky!

**Pull request merged**